### ISSUE TITLE
Assets fail to load in a multi-module project

### ISSUE THREAD

**Issue by mokun** [AlmasB/FXGL#425]
Hi, I run into the following problem when running the demo based on https://github.com/AlmasB/FXGL/wiki/Adding-Images-and-Sounds : 

> 12:03:59.131 [JavaFX Application Thread] FATAL AssetLoader          - Loading failed for asset: drop.wav. Cause: Unrecognized file signature!
> 12:03:59.131 [JavaFX Application Thread] FATAL GameApplication      - Uncaught Exception:
> 12:03:59.132 [JavaFX Application Thread] FATAL GameApplication      - 
> 
> Exception occurred: java.lang.IllegalArgumentException : Failed to load asset: drop.wav. Cause: Unrecognized file signature!
> E: com.almasb.fxgl.asset.AssetLoader.loadFailed(AssetLoader.kt:652)
> E: com.almasb.fxgl.asset.AssetLoader.loadSound(AssetLoader.kt:216)
> E: com.almasb.fxgl.audio.AudioPlayer.playSound(AudioPlayer.kt:114)
> E: FXGLBasicGameApp.lambda$0(FXGLBasicGameApp.java:103)

Both drop.wav and brick.png won't load.

In my project, I leave the demo class file within a maven package. The AssetLoader has trouble looking for my resources folder (let alone the textures and sounds within). 

I've tried using getResource() to get the relative path 

`String path = this.getClass().getResource("../../assets/textures/brick.png").toExternalForm();`

but AssetLoader doesn't understand it. 

Is there a way to load assets from resources in this case ? 

Thanks much for any insight and keep up the good work !

**Comment by AlmasB** [comment]
Hi, what's your directory structure?

**Comment by mokun** [comment]
I use Eclipse and have the following structure : 
![resources folder](https://user-images.githubusercontent.com/1168584/32693216-dd6c9076-c6db-11e7-951d-21f25374f367.png)

**Comment by AlmasB** [comment]
What about your pom.xml?

**Comment by mokun** [comment]
> What about your pom.xml?

I actually have multiple maven modules and each module has its own pom.xml. So it's very complicated :(

I load fxgl in maven as follows in one of the pom.xml : 

	<dependency>
		<groupId>com.github.almasb</groupId>
		<artifactId>fxgl</artifactId>
		<version>0.3.9</version>
	</dependency>

is pom.xml important in determining the resources path ?

**Comment by AlmasB** [comment]
Well, as long as resources are being copied properly, it should be fine. Now, I'm more concerned about how you start FXGL application, is it just a copy from the tutorial?

**Comment by AlmasB** [comment]
Also, it seems you are trying to integrate FXGL into something complex. Perhaps you should first do a standalone demo project, just to make sure it does what you want?

**Comment by mokun** [comment]
So if I have the following : 

```
    @Override
    protected void initGame() {
    	//String path = "brick.png";
    	String path = this.getClass().getResource("/assets/textures/brick.png").toExternalForm();
    	System.out.println("path : " + path);
        player = Entities.builder()
                .at(300, 300)
                .viewFromTexture(path)
                .buildAndAttach(getGameWorld());
	}

```

It would show that the "/assets/textures/" was inserted in front of the real path. 

> Exception occurred: java.lang.IllegalArgumentException : Failed to load asset: file:/D:/Data/git/mars-sim/mars-sim-ui/target/classes/assets/textures/brick.png. Cause: Asset "/assets/textures/file:/D:/Data/git/mars-sim/mars-sim-ui/target/classes/assets/textures/brick.png" was not found!

In my project setup, the real path turns out to be "D:/Data/git/mars-sim/mars-sim-ui/target/classes/assets/textures/brick.png"

In other places in my project, I can load my resources very easily as follows :  

`primaryStage.getIcons().add(new Image(this.getClass().getResource("/icons/lander_hab64.png").toExternalForm()));`

**Comment by AlmasB** [comment]
Ideally,
```
@Override
protected void initGame() {
    player = Entities.builder()
            .at(300, 300)
            //.viewFromNode(new Rectangle(25, 25, Color.BLUE))
            .viewFromTexture("brick.png")
            .buildAndAttach(getGameWorld());
}
```
should _just_ work given your directory structure.

You _can_ do it this way, if the above fails:
```
Image image = ...
Texture texture = new Texture(image);
```

Asset loader was designed so that all assets are self-contained and clearly defined, there wouldn't be a need for specifying the exact path since a texture is always under `assets/textures/`.

**Comment by AlmasB** [comment]
My best bet is that the resources are getting copied to some "global" project directory, and FXGL is searching a local project (pom.xml) directory, or vice versa.

**Comment by AlmasB** [comment]
I've created a multi-module project at [FXGL-MultiMaven](https://github.com/AlmasB/FXGL-MultiMaven) and will try to fix this by end of today.

**Comment by AlmasB** [comment]
I've just tested with a multi-module project both on Java 8 and 9. It works. Can you confirm that you are still having issues?

**Comment by mokun** [comment]
I created a brand new project but have the same issue of loading drop.wav. see below : 

![drop wav not loadable](https://user-images.githubusercontent.com/1168584/32694053-7da16d3c-c6eb-11e7-9fae-7a928198ea6e.png)

**Comment by AlmasB** [comment]
Can you check that `target/classes/assets/sounds/drop.wav` is present?

**Comment by AlmasB** [comment]
Also, does `brick.png` work?

**Comment by mokun** [comment]
yes both drop.wav and brick.png are there in the folder hierarchy in that screenshot. brick.png cannot be loaded to the screen when I run the class but the stack trace doesn't complain. It only complains about the drop.wav after I move > 100 pixels. don't know why.

I'll try your multi-module project at FXGL-MultiMaven now.

**Comment by mokun** [comment]
see my pom.xml on this new project : 

![drop wav path](https://user-images.githubusercontent.com/1168584/32694148-4c2d4e2c-c6ed-11e7-9c95-40c9133a8161.png)

**Comment by AlmasB** [comment]
I've also run it in Eclipse Oxygen.1a (4.7.1a), it loads just fine. Does the `AssetsSample` work from MultiMaven?

**Comment by AlmasB** [comment]
Just some thoughts... `Unrecognized file signature!` is a JavaFX media exception, I wonder if you are missing some codecs or something, but `brick.png` should work regardless of that and should display a brick texture on the screen.

**Comment by AlmasB** [comment]
I've got another idea. Download and run [this game](https://github.com/AlmasB/FXGLGames/blob/master/binaries/SpaceInvaders-1.0.jar). It was generated by the FXGL build automatically. If the game runs and plays with sounds / textures, then there is something odd with your IDE / build. If the game fails (see under `logs/`) then something's wrong with the environment.

**Comment by mokun** [comment]
Your FXGL-MultiMaven works in my Eclipse environment. 

So this comes down to the way I set up my project was somehow incorrect.

Could it be simply that my pom.xml is missing something ?

> <?xml version="1.0" encoding="UTF-8"?>
> <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
> 	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
> 	<modelVersion>4.0.0</modelVersion>
> 	<groupId>io.github.mars-sim</groupId>
> 	<artifactId>mars-sim</artifactId>
> 	<version>3.1.1-SNAPSHOT</version>
> 
>     <properties>
>         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
>         <source.version>1.8</source.version>
>         <maven.compiler.version>3.7.0</maven.compiler.version>
>     </properties>
>     
> 	<dependencies>
> 		<dependency>
> 			<groupId>com.github.almasb</groupId>
> 			<artifactId>fxgl</artifactId>
> 			<version>0.3.9</version>
> 		</dependency>
> 	</dependencies>
> 
> 	<build>
>         <plugins>
>             
>             <plugin>
>                 <artifactId>maven-compiler-plugin</artifactId>
>                 <version>${maven.compiler.version}</version>
>                 <configuration>
>                     <source>${source.version}</source>
>                     <target>${source.version}</target>
>                 </configuration>
>             </plugin>
>         </plugins>
>     </build>
>     
> </project>

**Comment by mokun** [comment]
Yes your AssetsSample class works within your project package. As soon as I move it to my project package, it doesn't work. 

Still, I wonder what I'm missing in setting up this brand new maven project for running FXGL ?

**Comment by AlmasB** [comment]
[This](https://github.com/AlmasB/FXGL-Maven) is the minimal Maven project, there shouldn't be anything extra that you need to add, so I have no idea at the moment...

**Comment by AlmasB** [comment]
Can you maybe line by line change the project that works (FXGL-MultiMaven?) to become what you need? If you change 1 line at a time and keep running it, maybe you'll find out what's causing the issue.

**Comment by mokun** [comment]
Since your minimal maven project doesn't have the resources folder, I have to create it myself and copy and paste the assets folder/subfolders over. It still didn't work and complained about "Unrecognized file signature". So, I delete the assets folders and copy and paste again one or two times, suddenly it works !?

**Comment by mokun** [comment]
If I deleted the assets folder and its subfolders and run the class, the stack trace will say that it can't find the brick.png and drop.wav, which is correct.

> 15:54:31.904 [FXGL Background Thread 1 ] FATAL AssetLoader          - Loading failed for asset: brick.png. Cause: Asset "/assets/textures/brick.png" was not found!
> 15:54:31.924 [JavaFX Application Thread] FATAL GameApplication      - Uncaught Exception:
> 15:54:31.925 [JavaFX Application Thread] FATAL GameApplication      - 
> 
> Exception occurred: java.lang.IllegalArgumentException : Failed to load asset: brick.png. Cause: Asset "/assets/textures/brick.png" was not found!

So "Unrecognized file signature" is a totally different animal that we're dealing with.

**Comment by mokun** [comment]
Now It works in both my mars-sim project and my brand new maven project. 

Also, when I created a new maven project from the scratch, what I did was that I manually created a empty folder called "assets" and the subfolders "textures" and "sounds". and put drop.wav and brick.png in them. I did the same thing manually on my mar-sim project. 

So I delete the assets folder and then proceed to copy and paste the assets folder hierarchy from your FXGL-MultiMaven project over to my new maven project, and VIOLA it works ! 

So copy and paste makes it works somehow. 

I figure it's just some sort of peculiarity of eclipse when it comes to configuring the Build Path of the resources folder :(
